tempesta_fw/sched/tfw_sched_hash.c

```diff
@@ -245,11 +200,105 @@ tfw_sched_hash_get_srv_conn(TfwMsg *msg, TfwServer *srv)
 	return NULL;
 }
 
+static void
+tfw_sched_hash_cleanup(TfwSrvGroup *sg)
+{
+	size_t si;
+	TfwHashSrvList *sl = sg->sched_data;
+
+	if (!sl)
+		return;
+
+	for (si = 0; si < sl->srv_n; ++si) {
+		if (sl->srvs[si].conn)
+			kfree(sl->srvs[si].conn);
+		if (sl->srvs[si].hash)
+			kfree(sl->srvs[si].hash);
+	}
+
+	kfree(sl);
+	sg->sched_data = NULL;
+}
+
+static void
+tfw_sched_hash_del_grp(TfwSrvGroup *sg)
+{
+	tfw_sched_hash_cleanup(sg);
+}
+
+static int
+tfw_sched_hash_add_grp(TfwSrvGroup *sg)
+{
+	int ret = -ENOMEM;
+	size_t size, si, ci;
+	unsigned int sum_conn_n;
+	TfwServer *srv;
+	TfwSrvConn *srv_conn;
+	TfwHashSrv *hsrv;
+	TfwHashSrvList *sl;
+
+	/*
+	 * Validate the number of servers in the group, and the number
+	 * of connections for each server.
+	 */
+	si = 0;
+	list_for_each_entry(srv, &sg->srv_list, list) {
```

Comment:
I'd rather leave the validation check. I've moved it to a separate function. It's called only once for each group at configuration processing time, and I consider it a good safety measure.

As for the comment on the new behaviour, it's here: https://github.com/tempesta-tech/tempesta/blob/bab5e182a4d4192a308e1510600832d5bdceb905/tempesta_fw/server.h#L116
I have extended it a little bit to be more visible and clear.